File: src/api/ws/connection_manager.py
import asyncio
from typing import Dict, List, Set

import structlog
from starlette.websockets import WebSocket

logger = structlog.get_logger("ws.connection_manager")


class ConnectionManager:
    """
    Tracks all active WebSocket connections and their topic subscriptions.
    All lookups are O(1) dict operations. Memory per connection: ~4 KB.
    """

    __slots__ = (
        "_connections",
        "_subscriptions",
        "_topic_subscribers",
        "_client_scopes",
    )

    def __init__(self) -> None:
        self._connections: Dict[str, WebSocket] = {}  # client_id → WebSocket
        self._subscriptions: Dict[str, Set[str]] = {}  # client_id → {topic}
        self._topic_subscribers: Dict[str, Set[str]] = {}  # topic → {client_id}
        self._client_scopes: Dict[str, dict] = {}  # client_id → {db_scope, fs_scope}

    async def connect(self, websocket, client_id: str, scopes: dict) -> None:
        await websocket.accept()
        self._connections[client_id] = websocket
        self._subscriptions[client_id] = set()
        self._client_scopes[client_id] = scopes
        logger.info("WebSocket connected", client_id=client_id)

    def disconnect(self, client_id: str) -> None:
        for topic in self._subscriptions.pop(client_id, set()):
            self._topic_subscribers.get(topic, set()).discard(client_id)
        self._connections.pop(client_id, None)
        self._client_scopes.pop(client_id, None)
        logger.info("WebSocket disconnected", client_id=client_id)

    def subscribe(self, client_id: str, topic: str) -> bool:
        """Returns True if subscribed, False if topic is outside the key's scope."""
        if not self._topic_in_scope(topic, self._client_scopes.get(client_id, {})):
            return False
        self._subscriptions.setdefault(client_id, set()).add(topic)
        self._topic_subscribers.setdefault(topic, set()).add(client_id)
        return True

    def unsubscribe(self, client_id: str, topic: str) -> None:
        self._subscriptions.get(client_id, set()).discard(topic)
        self._topic_subscribers.get(topic, set()).discard(client_id)

    def get_subscribers(self, topic: str) -> List[str]:
        return list(self._topic_subscribers.get(topic, set()))

    async def send(self, client_id: str, payload: bytes) -> None:
        ws = self._connections.get(client_id)
        if ws is None:
            return
        try:
            await ws.send_bytes(payload)
        except Exception:
            self.disconnect(client_id)

    async def broadcast(self, topic: str, payload: bytes) -> None:
        """Broadcast a pre-serialized orjson payload to all topic subscribers."""
        tasks = [self.send(cid, payload) for cid in self.get_subscribers(topic)]
        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)

    def _topic_in_scope(self, topic: str, scopes: dict) -> bool:
        parts = topic.split(":", 2)
        if not parts:
            return False
        module = parts[0]
        if module == "db":
            alias = parts[1] if len(parts) > 1 else ""
            db_scope = scopes.get("db_scope", [])
            return "*" in db_scope or alias in db_scope
        if module == "fs":
            alias = parts[1] if len(parts) > 1 else ""
            fs_scope = scopes.get("fs_scope", [])
            return "*" in fs_scope or alias in fs_scope
        # metrics / system topics are open to all authenticated clients
        return True

    @property
    def active_count(self) -> int:
        return len(self._connections)

    @property
    def topic_count(self) -> int:
        return len(self._topic_subscribers)


# Module-level singleton — shared across all requests in the same process
conn_mgr = ConnectionManager()

File: src/api/ws/connection_manager.rs
use crate::config::loader::ConfigManager;
use axum::extract::ws::Message;
use once_cell::sync::Lazy;
use std::collections::{HashMap, HashSet};
use tokio::sync::mpsc;
use tokio::sync::RwLock;

#[derive(Clone, Default)]
pub struct ClientScopes {
    pub db_scope: Vec<String>,
    pub fs_scope: Vec<String>,
}

pub struct ConnectionManager {
    connections: RwLock<HashMap<String, mpsc::UnboundedSender<Message>>>,
    subscriptions: RwLock<HashMap<String, HashSet<String>>>,
    topic_subscribers: RwLock<HashMap<String, HashSet<String>>>,
    client_scopes: RwLock<HashMap<String, ClientScopes>>,
}

impl ConnectionManager {
    pub fn new() -> Self {
        Self {
            connections: RwLock::new(HashMap::new()),
            subscriptions: RwLock::new(HashMap::new()),
            topic_subscribers: RwLock::new(HashMap::new()),
            client_scopes: RwLock::new(HashMap::new()),
        }
    }

    pub async fn register(
        &self,
        client_id: &str,
        sender: mpsc::UnboundedSender<Message>,
        scopes: ClientScopes,
    ) {
        self.connections
            .write()
            .await
            .insert(client_id.to_string(), sender);
        self.subscriptions
            .write()
            .await
            .insert(client_id.to_string(), HashSet::new());
        self.client_scopes
            .write()
            .await
            .insert(client_id.to_string(), scopes);
        println!("WebSocket connected: {}", client_id);
    }

    pub async fn disconnect(&self, client_id: &str) {
        let subs = self
            .subscriptions
            .write()
            .await
            .remove(client_id)
            .unwrap_or_default();
        let mut topic_subs = self.topic_subscribers.write().await;
        for topic in subs {
            if let Some(clients) = topic_subs.get_mut(&topic) {
                clients.remove(client_id);
            }
        }
        self.connections.write().await.remove(client_id);
        self.client_scopes.write().await.remove(client_id);
        println!("WebSocket disconnected: {}", client_id);
    }

    pub async fn subscribe(&self, client_id: &str, topic: &str) -> bool {
        let config = ConfigManager::get();
        let max_subs = config.websocket.max_subscriptions_per_client as usize;

        let mut subs_map = self.subscriptions.write().await;
        let subs = subs_map.entry(client_id.to_string()).or_default();

        if !subs.contains(topic) && subs.len() >= max_subs {
            return false;
        }

        let scopes_map = self.client_scopes.read().await;
        let scopes = scopes_map.get(client_id).cloned().unwrap_or_default();

        if !self.topic_in_scope(topic, &scopes) {
            return false;
        }

        subs.insert(topic.to_string());
        drop(subs_map); // unlock before locking next

        self.topic_subscribers
            .write()
            .await
            .entry(topic.to_string())
            .or_default()
            .insert(client_id.to_string());

        true
    }

    pub async fn unsubscribe(&self, client_id: &str, topic: &str) {
        if let Some(subs) = self.subscriptions.write().await.get_mut(client_id) {
            subs.remove(topic);
        }
        if let Some(clients) = self.topic_subscribers.write().await.get_mut(topic) {
            clients.remove(client_id);
        }
    }

    pub async fn send(&self, client_id: &str, payload: String) {
        let connections = self.connections.read().await;
        if let Some(sender) = connections.get(client_id) {
            let _ = sender.send(Message::Text(payload));
        }
    }

    pub async fn broadcast(&self, topic: &str, payload: String) {
        let topic_subs = self.topic_subscribers.read().await;
        let connections = self.connections.read().await;

        if let Some(subscribers) = topic_subs.get(topic) {
            for cid in subscribers {
                if let Some(sender) = connections.get(cid) {
                    let _ = sender.send(Message::Text(payload.clone()));
                }
            }
        }
    }

    fn topic_in_scope(&self, topic: &str, scopes: &ClientScopes) -> bool {
        let parts: Vec<&str> = topic.splitn(3, ':').collect();
        if parts.is_empty() {
            return false;
        }

        let module = parts[0];
        if module == "db" {
            let alias = if parts.len() > 1 { parts[1] } else { "" };
            return scopes.db_scope.iter().any(|s| s == "*" || s == alias);
        }
        if module == "fs" {
            let alias = if parts.len() > 1 { parts[1] } else { "" };
            return scopes.fs_scope.iter().any(|s| s == "*" || s == alias);
        }
        true
    }

    pub async fn active_count(&self) -> usize {
        self.connections.read().await.len()
    }
}

pub static CONN_MGR: Lazy<ConnectionManager> = Lazy::new(ConnectionManager::new);
